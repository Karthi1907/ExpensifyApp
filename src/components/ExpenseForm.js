import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';


const now = moment();
console.log(now.format('YYYY MMMM Do'));   
console.log(now.startOf('month').format('YYYYMMDD'));


export default class ExpenseForm extends React.Component {

    constructor(props) {
        super(props);
        this.state = {
            descr: props.expense ? props.expense.descr : '',
            comments: props.expense ? props.expense.comments : '',
            amount: props.expense ? props.expense.amount : '0.00',
            createdAt: props.expense ? moment(props.expense.createdAt, 'YYYYMMDD')   : moment(),
            calendarFocused: false,
            error: undefined
        }
    }



    onDescriptionChange = (e) => {

        const description = e.target.value;
        this.setState( () => ({ descr: description }) );
        console.log(this.state);
    };

    onCommentsChange = ( e ) => {
        //const comments = e.target.value;
        e.persist();
        this.setState( () => ( { comments: e.target.value }) );
        console.log(this.state);
    }

    onAmountChange = (e) => {
        const amount = e.target.value;
        if (!amount ||  amount.match(/^\d{1,7}(\.\d{0,2})?$/) ) {
            this.setState( () => ({  amount }) );
        }
    }

    onDateChange = ( createdAt ) => {
        
        if (createdAt) {
            const createdAtDate = createdAt.format('YYYYMMDD');
            console.log(createdAtDate);
            this.setState( () => ( { createdAt } ) );
            console.log(createdAt.format('YYYYMMDD'));
            console.log(this.state);    
            console.log(this.state.createdAt);
        }
    }

    onFocusChange = ( {focused} ) => {
        this.setState( () => ( { calendarFocused: focused } ) );
        console.log(this.state);
    }

    onSubmit = (e) => {

        e.preventDefault();
        console.log('Inside onSubmit() function');
        let error = '';

        if ( !this.state.descr || !this.state.amount ) {
            error = 'Mandatory inputs missing - add both description and amount to continue';
        }

        if (error) {
            this.setState( () => ({ error: error }) );
        } else {
            this.setState( () => ({ error: '' }) );
        
            console.log(error);
            console.log(this.state); 
            this.props.onSubmit( {
                descr: this.state.descr,
                amount: parseFloat(this.state.amount, 10),
                createdAt: this.state.createdAt.format('YYYYMMDD'),
                comments: this.state.comments
            } ) ;  
        }
    };


    render()
    {
        return (
            <form onSubmit={this.onSubmit}> 
            { this.state.error && <p style={{color: 'red',fontWeight:'bold'}}>  {this.state.error} </p> }   
            <div>
                <div>
                    <p> Description: </p>
                    <input 
                        type="text"
                        placeholder="Your Expense Descr"
                        autoFocus
                        value={this.state.descr}
                        onChange={this.onDescriptionChange}
                    />
                </div>
                <div>
                    <p> Amount:</p>
                    <input 
                        type="text"
                        value={this.state.amount}
                        placeholder="Your Expense Amount"
                        onChange={this.onAmountChange}
                    />
                </div>
                <div>
                <p> Create Date: </p>
                <SingleDatePicker 
                    date={this.state.createdAt}
                    onDateChange={this.onDateChange}
                    focused={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    displayFormat="DD-MM-YYYY"
                    numberOfMonths={1}
                    isOutsideRange={ () => false }
                />
                </div>
                <div>
                    <p> Comments</p>
                    <textarea
                        type="text"
                        placeholder="Add Comments for your expense here"
                        value={this.state.comments}
                        onChange={this.onCommentsChange}
                    />
                </div>
                <div> 
                    <p> </p>
                    <button> Add Expense </button>
                </div> 
            </div>
            </form>
        );
    }
};