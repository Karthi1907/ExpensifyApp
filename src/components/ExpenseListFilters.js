import React from 'react';
import { connect } from 'react-redux';
import {setFilterText, sortByDate, sortByAmount, setAscendingOrder, setDescendingOrder, setStartDate, setEndDate} from '../actions/filters';
import { DateRangePicker } from 'react-dates';

export class ExpenseListFilters extends React.Component {

    state = {
        calendarFocusedInput : null
    };

    onDatesChange = ( {startDate, endDate} ) => {
        this.props.setStartDate ( startDate ) ;
        this.props.setEndDate( endDate  );
    };

    onFocusChange = (calendarFocusedInput) => {
        this.setState( () => ({ calendarFocusedInput }) ); 
    };
    
    onFilterTextChange = ( e ) => {
        this.props.setFilterText(e.target.value);
    };

    onSortBySelectionChange = ( e ) => {
        console.log(e.target.value);
        e.target.value === 'date' ? this.props.sortByDate() : this.props.sortByAmount();
    };

    onSortOrderChange = ( e ) => {
        console.log(e.target.value);
        e.target.value === "asc" ? this.props.setAscendingOrder() : this.props.setDescendingOrder() ;                   
    }


    render() {
        return (
            <div>
            <input type = "text" value={this.props.filters.text} onChange={ this.onFilterTextChange} />
            <select value={this.props.filters.sortBy} onChange={ this.onSortBySelectionChange }>
                <option value="amount"> Amount </option>
                <option value="date"> Date </option>
            </select> 
    
            <select value={this.props.filters.order}
                onChange={ this.onSortOrderChange } >
                    <option value="asc">  Ascending Order </option>
                    <option value="desc"> Descending Order </option>
            </select>
            <DateRangePicker 
                startDate={this.props.filters.startDate}
                startDateId='startDateId'
                endDate={this.props.filters.endDate}
                endDateId='endDateId'
                onDatesChange={ this.onDatesChange }
                focusedInput={ this.state.calendarFocusedInput }
                onFocusChange={ this.onFocusChange }
                numberOfMonths={1}
                isOutsideRange={() => false}
                
                showClearDates={true}
                

            />
        </div>


        )
    }


}

// const ExpenseListFilters = ( props ) => (
//     <div>
//         <input type = "text" value={props.filters.text} onChange={ (e) => {
//             props.dispatch(setFilterText(e.target.value));
//         } } />
//         <select value={props.filters.sortBy} onChange={ (e) => {
//             console.log(e.target.value);
//             e.target.value === 'date' ? props.dispatch(sortByDate()) : props.dispatch(sortByAmount());
//         } }>
//             <option value="amount"> Amount </option>
//             <option value="date"> Date </option>
//         </select> 

//         <select value={props.filters.order}
//             onChange={(e) => {
//                 console.log(e.target.value);
//                 e.target.value === "asc" 
//                     ? props.dispatch( setAscendingOrder() ) 
//                     : props.dispatch ( setDescendingOrder() );                   
//             } } >
//                 <option value="asc">  Ascending Order </option>
//                 <option value="desc"> Descending Order </option>
//         </select>
//         <DateRangePicker 
//             startDate={props.filters.startDate}
//             endDate={}
        
//         />

    

//     </div>
// );

const mapStateToProps = (state) => {

    return {
       // expenses: state.expenses,
        filters: state.filters
    };

};

const mapDispatchToProps = (dispatch) => {
    return {
        setStartDate: (startDate) => dispatch(setStartDate(startDate)),
        setEndDate: (endDate) => dispatch(setEndDate(endDate)),
        setFilterText: (textVal) => dispatch(setFilterText(textVal)),
        sortByDate: () => dispatch(sortByDate()),
        sortByAmount: () => dispatch(sortByAmount()),
        setAscendingOrder: () => dispatch(setAscendingOrder()),
        setDescendingOrder: () => dispatch(setDescendingOrder())
    }
};

const ConnectedExpenseListFilters = connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);

export default ConnectedExpenseListFilters;