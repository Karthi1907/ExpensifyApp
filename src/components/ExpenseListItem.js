import React from 'react';
import { connect } from 'react-redux';
import { removeExpense } from '../actions/expenses';
import { Link } from 'react-router-dom';

export const ExpenseListItem = ( {dispatch, id, descr, amount,  createdAt } ) => (

    <div>
        <Link to={`/edit/${id}`} >
            <h3>  {descr}    </h3>
        </Link>
        <p>  {amount} - {createdAt} </p> 
        <button onClick={(e) => {
            dispatch( removeExpense(id) );
        }}>Remove me </button>
        
    </div>

);

const ConnectedExpenseListItem = connect()(ExpenseListItem);

export default ConnectedExpenseListItem;

