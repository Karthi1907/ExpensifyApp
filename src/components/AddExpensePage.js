import React from 'react';
import {connect} from 'react-redux';
import { Link } from 'react-router-dom';
import ExpenseForm from './ExpenseForm';
import { addExpense } from '../actions/expenses';

/* Version 1 - Below version is difficult to test via TestSpies; hence changing to a version using mapDispatchToProps */

// const AddExpensePage = (props) => (
//     <div>
//         <ExpenseForm 
//           onSubmit={ ( expense ) => {
//             props.dispatch( addExpense(expense) );
//             props.history.push('/');
//           }} />
//       <Link to="/"> Go Home </Link>
//     </div>
//   );

// const ConnectedAddExpensePage = connect()(AddExpensePage);

/* Version 2 - Changes to use mapDispatchToProps */

// const AddExpensePage = (props) => (
//   <div>
//     <h1> Add Expense Form </h1>
//     <ExpenseForm onSubmit={ (expense) => {
//           props.onSubmit(expense);
//           props.history.push('/');        
//         } 
//       }
//     />
//   </div>
// );

/* Version 3 - AddExpensePage can be changed into a class based component so that we no longer have to define the 
  onSubmit() function inline. */

export class AddExpensePage extends React.Component {
  
  onSubmit = ( expense ) => {
    this.props.onSubmit(expense);
    this.props.history.push('/');
  }
 
  render() {
    return (
    <div>
      <h1> Add Expense Form </h1>
        <ExpenseForm onSubmit={this.onSubmit} />
      </div>
    );
  }
}

const mapDipatchToProps = ( dispatch ) => ( { onSubmit: (expense) => dispatch(addExpense(expense) ) } );

const ConnectedAddExpensePage = connect(undefined, mapDipatchToProps)(AddExpensePage);

export default ConnectedAddExpensePage;