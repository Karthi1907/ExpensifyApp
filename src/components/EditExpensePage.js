import React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense } from '../actions/expenses';
import { removeExpense } from '../actions/expenses'; 

// const EditExpensePage = (props) => 
// (
//     <div>
//       <ExpenseForm 
//         expense={props.expense}
//         onSubmit={ (expense) => {
//             props.dispatch( editExpense ( props.expense.id , {...expense} ) );
//             props.history.push('/');
//         } }
//       />
//       <button onClick={ (e) => {
//            props.dispatch( removeExpense (props.expense.id) );
//            props.history.push('/');       
//       }}>
//           Remove
//       </button>
//       <Link to="/"> Go Home </Link>
//       {/* <p> { props.expense.descr } </p> */}
//     </div>
//   );

export class EditExpensePage extends React.Component {

  editExpense = ( expense ) => {
    this.props.editExpense(this.props.expense.id, expense);
    this.props.history.push('/');
  };

  removeExpense = ( e ) => {
    e.preventDefault();
    this.props.removeExpense( this.props.expense.id );
    this.props.history.push('/');
  };

  render() {
    return (
     <div>
       <ExpenseForm  expense={this.props.expense} onSubmit={ this.editExpense }
       />
       <button onClick={ this.removeExpense }> Remove </button>
       <Link to="/"> Go Home </Link>
       {/* <p> { props.expense.descr } </p> */}
     </div>      
    );
  };
};


const mapStateToProps = ( state , props ) => {
    return {
      expense: state.expenses.find( (expense) => expense.id === props.match.params.id )
    };
};

const mapDispatchToProps = ( dispatch ) => {
  return {
    editExpense: (id, expense ) => dispatch(editExpense(  id, {...expense}) ),
    removeExpense: ( id) => dispatch(removeExpense( id )) 
  }
} 

const ConnectedExoensePage = connect(mapStateToProps, mapDispatchToProps)( EditExpensePage );  

export default ConnectedExoensePage;     