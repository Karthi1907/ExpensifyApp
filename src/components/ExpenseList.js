import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import getExpenses from '../selectors/expenses';

export const ExpenseList = (props) => (

    <div>
        <h3>
            Expense List
        </h3>
        <p> Matching Expenses: {props.expenses.length} </p>
        {/* <p> {props.filters.text} </p> */}
        {/* { props.expenses.map( (expense)  => 
            {
                return ( <ExpenseListItem key={expense.id} {...expense} /> );
            } ) 
        }; */}

        { props.expenses.length === 0   
            ? ( <p>  No expenses to show </p> )  
            : ( props.expenses.map( (expense)  => 
                <ExpenseListItem key={expense.id} {...expense} /> ) )
        }

        {/* { props.expenses.map( (expense)  =>  <ExpenseListItem key={expense.id} {...expense} /> )}; */}
           
        
    </div>
);

const mapStateToProps = (state) => {
    return {
        // expenses : state.expenses, filters : state.filters
        expenses: getExpenses(state.expenses, state.filters)

    }
};

const ConnectedExpenseList = connect(mapStateToProps)(ExpenseList);

// const ConnectedExpenseList = connect( (state) => {
//     return {
//         expenses: state.expenses,
//         filters: state.filters
//     };
// } )(ExpenseList);

export default ConnectedExpenseList;

