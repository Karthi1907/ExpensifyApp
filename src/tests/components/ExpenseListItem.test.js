import React from 'react';
import {ExpenseListItem} from '../../components/ExpenseListItem';
import {shallow} from 'enzyme';

test('should render ExpenseListItem component properly', () => {
    const expense = {
        dipatch: {type: 'ADD_EVENT'},
        id: 10,
        edscr: 'Description',
        amount: 123.45,
        createdAt: 20281001
    }
    const wrapper = shallow( 
        <ExpenseListItem {...expense}/> );
    expect(wrapper).toMatchSnapshot();
} );