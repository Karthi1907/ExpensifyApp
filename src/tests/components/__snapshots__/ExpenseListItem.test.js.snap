// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseListItem component properly 1`] = `
<div>
  <Link
    replace={false}
    to="/edit/10"
  >
    <h3>
        
          
    </h3>
  </Link>
  <p>
      
    123.45
     - 
    20281001
     
  </p>
  <button
    onClick={[Function]}
  >
    Remove me 
  </button>
</div>
`;
